core/process.py

```diff
@@ -494,8 +495,16 @@ def Exec(self, arg_vec, environ):
     try:
       os_.execvpe(argv[0], argv, environ)
     except OSError as e:
-      # TODO: Run with /bin/sh when ENOEXEC error (noshebang).  Because all
-      # shells do it.
+      # Run with /bin/sh when ENOEXEC error (no shebang). Because all shells do it.
+      if e.errno == errno.ENOEXEC:
+        _, realpath = _ResolveFile(argv[0], environ["PATH"].split(':'))
+        # Check if file was deleted in the meantime
+        if realpath is None:
+          e.errno = errno.ENOENT
+        else:
+          arg_vec.strs = ["/bin/sh", realpath] + argv[1:]
```

Comment:
I guess this is technically correct, but I don't like the style of mutating arg_vec either... it breaks the invariant that arg_vec.strs and arg_vec.spids are parallel (and the same length).

Reply:
How would I get an spid to add to arg_vec? `Exec` uses it on errors.

I guess that brings up a second question of if the original fails _and_ `/bin/sh` fails whether we want to print '/bin/sh' or the original, that's kind of a weird edge case.

Reply:
One way is to use const.NO_INTEGER, e.g.

arg_vector(["/bin/sh" ... , [const.NO_INTEGER] + arg_vec.spids)

Or another way is factoring out a private function like _Exec(self, argv, argv0_spid), so there's no arg_vector any more.

-----

I think this code also will enter an infinite loop if `/bin/sh` returns ENOEXEC.  That is a corner case but we should probably avoid these things even in theory, since shells are often used to fix broken systems, etc.


------------

Note this code may change in the near future due to issue #355, which caches the PATH lookups.